core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -172,6 +175,132 @@ void *CudaSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
+
+/**
+ * If a device has been initialized we will drop it in the set
+ * so we don't initialize it again
+ */
+std::unordered_set<int> _mem_pool_initialized;
+
+/**
+ * This is the initializer. It currently gets the device_id and stream
+ * that impl_alloc gets
+ *    @param device_id  The device id on which to alter the default mempool
+ *    @param stream     A specific stream for cudaMallocAsync
+ * 		@param error_code return any failure codes for cuda APIs
+ *    @return  				  true for success, false for failure
+ */
+bool initializeMempool(const int device_id, const cudaStream_t stream,
+                       cudaError_t &error_code) {
+  error_code = cudaSuccess;
+
+  std::cout << "Initializing Default Memory Pool for device " << device_id
+            << "\n";
+
+  // Interpret env var
+  char *env_string = getenv("KOKKOS_CUDA_MEMPOOL_SIZE");
+  if (!env_string) return true;  // Nothing to set, initialization is complete
+
+  // Env var can be a decimal e.g. 2.4g
+  double factor = 1;  // bytes
+
+  // Interpret the string
+  std::string mempool_size_string(env_string);
+  char last_letter = mempool_size_string.back();
+
+  // We can deal with the last letter being units.
+  if (std::isalpha(last_letter)) {
+    mempool_size_string.pop_back();
+    char last_letter_cap = std::toupper(last_letter);
+    switch (last_letter_cap) {
+      case 'K': factor = 1024; break;
+      case 'M': factor = 1024 * 1024; break;
+      case 'G': factor = 1024 * 1024 * 1024; break;
+      default:
+        std::cerr << "Unknown units for requested size " << last_letter << "\n";
+        return false;
+        break;
+    }
+  }
+
+  // At this point we removed the last letter (or there was not a last letter)
+  // convert what is left to a number.
+  double requested_size = 0;
+
+  // Handle exception in case the string is unconvertible
+  try {
+    requested_size = static_cast<size_t>(std::stod(mempool_size_string));
+  } catch (...) {
+    std::cerr << "Unable to convert " << mempool_size_string
+              << " to a number\n";
+    return false;
+  }
+
+  // Check for non-positive size memory requests
+  if (requested_size <= 0) {
+    std::cerr << "Negative amount of memory requested in allocation\n";
+    return false;
+  }
+
+  // Multiply in units
+  requested_size *= factor;
+
+  // Check we are not asking for memory that is greater than what size_t can
+  // hold. Since requested can be the larger I convert the maximum of size_t
+  // to a double and compare those
+  double max_size_t = static_cast<double>(std::numeric_limits<size_t>::max());
+  if (requested_size > max_size_t) {
+    std::cerr << "Requested amount of memory " << requested_size << " exceeds "
+              << "maximum alloatable size " << max_size_t << "\n";
+    return false;
+  }
+
+  // At this point requested_size should be appropriate
+  // neither too big nor negative.
+  size_t n_bytes = static_cast<size_t>(std::ceil(requested_size));
```

Comment:
This could be larger than the largest representable `size_t`, though.

Reply:
On the other hand if I floor it we may be short. I can set the check on requested size to be   < max